lib/http/request/writer.rb

```diff
@@ -65,24 +68,35 @@ def send_request_header
         add_headers
         add_body_type_headers
 
-        @socket << join_headers
+        write(join_headers)
       end
 
       def send_request_body
         if @body.is_a?(String)
-          @socket << @body
+          write(@body)
         elsif @body.is_a?(Enumerable)
           @body.each do |chunk|
-            @socket << chunk.bytesize.to_s(16) << CRLF
-            @socket << chunk << CRLF
+            write(chunk.bytesize.to_s(16) << CRLF)
+            write(chunk << CRLF)
           end
 
-          @socket << ZERO << CRLF << CRLF
+          write(CHUNKED_END)
         end
       end
 
       private
 
+      def write(data)
+        while data.present?
+          length = @socket.write(data)
+          if data.length > length
+            data = data[length..-1]
```

Comment:
This will allocate every time. I wonder if it would be possible to use `#slice!` here to reduce allocations. Or maybe we just need real `ByteBuffers` :confused: